### DISCUSSION TITLE
How to Distinguish service name between multiple services running on a single JVM?

### DISCUSSION THREAD

**Discussion by rahjesh-vunet** [open-telemetry/opentelemetry-java-instrumentation#11986] [Q&A]
I have multiple services (war files) running in a JBoss server. As Opentelemetry Java agent is instrumenting the JVM there is no way to distinguish the traces from different services.
Referred these links.
https://github.com/open-telemetry/opentelemetry-java-instrumentation/discussions/10672
https://github.com/open-telemetry/opentelemetry-java-instrumentation/discussions/10537
I tried the following.
1.In the OTel collector, parse the span.attributes.http.route and get the context name and use it as service. But this works only for the root spans. The children spans don't have span.attributes.http.route. This ends up having rott span with correct service name and the other spans with "unknown_service:java"
2.We are storing the traces data in CH. Used the SQL to get the service name from the span.attributes.http.route. But here also it was very difficult to traverse the unknown level of parents to reach the root span.
Do you have any suggestion?

**Comment by laurit** [comment]
1. I have no idea how things work in the collector. I presume that if you want to tag all the spans in trace based on a root span attribute you'd need to build something similar to the [tail sampling processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/tailsamplingprocessor/README.md) that waits (or at least tries to) for the whole trace to arrive before making a decision.
2. I have no idea what CH is. I guess you are building a custom observability backend. You could check what the open source backends are doing. I'd assume (no, I have not looked at any) do some sort of post processing on the trace when it is complete.
3. Otel java agent offers an [extension](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/examples) mechanism that you could use to add attributes to spans. Have a look at https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/examples/extension/src/main/java/com/example/javaagent/DemoSpanProcessor.java Span processor gets called on span start. If it is a http server root span then you could use `url.path` attribute to figure out what the context path is and translate this to the application name. If it is a child span then you could look up the root span with https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/2af49b4c044db526dd0741eccfe0bdc07baeda71/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/LocalRootSpan.java#L35 (unfortunately you may need to cast it `ReadableSpan` to read attributes). Or you could write instrumentation that store the request context path or whatever you need in `context` then you could read the stored value in your span processor.
